PoX-4 Signer Set Structs, Getters & Setters [wip]

setzeus opened the pull request:
# PoX-4 Signer Set Structs, Getters, & Setters
## Summary of Changes

This PR is WIP addition to #4092 that addresses the PoX-4/Argon epic #3970 with the specific issue #4164. As titled, this specific PR introduces the data structures, getters & setters for the current & next-cycle signer set. 

Each cycle signer set can hold up to 4096 signers (identified by a (buff 33)), each with a list of 4096 key-ids (represented as a uint). The following functions were introduced:

1. Update-current-signer
2. Update-current-signer-set
3. Get-signer-set-by-reward-cycle
4. Get-signer-key-ids-by-reward-cycle

## Testing
### How were these changes tested?

This was only tested locally for instantiation. Unit tests will be address in #4092 prior to merging into Next.

### What future testing should occur?
The logic here should have corresponding unit tests in both Rust & Clarinet/Clarity. I'll be working with @moodmosaic to address these tests correctly.

## Checklist:
- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

Commit:
stashing demo changes

Commit:
mockamoto mining working with merge

setzeus committed:
updated stacking-state & added reward-cycle-signing-keys

setzeus committed:
updated delegate-stack-stx, stack-extend & delegate-stack-extend

setzeus committed:
updated stack-stx

Commit:
working integration with chains coordinator

Commit:
add test for mockamoto, cleanup implementation

Commit:
add transfer tx to mockamoto test, address PR feedback

setzeus committed:
comment updates

setzeus committed:
changed unwrap to unwrap-panic

Commit:
add stacker-self-signer to mockamoto, add boot receipts, add pox_v3_unlock to /new_block events

setzeus committed:
added simple instantiation test

setzeus committed:
Merge remote-tracking branch 'refs/remotes/origin/next' into feat/pox-4-signing-keys

setzeus committed:
possible formatting issue CI didn't like

setzeus committed:
ran cargo fmt --all for formatting CI issue

soju-drinker committed:
remove sbtc ops

setzeus committed:
Merge remote-tracking branch 'refs/remotes/origin/next' into feat/pox-4-signing-keys

setzeus committed:
updated with getter get-sgner-info-for-cycle

soju-drinker committed:
removing schemas

jbencin committed:
fix: Use `VRFProof::consensus_serialize()` to address issue #4115

setzeus committed:
Fixed failing unit tests by updating 'stacks-stx' parameters in existing tests

setzeus committed:
removed unnecessary signer-key optional wrapper

jbencin committed:
Fix unit tests and `TransactionPayload::deserialize()`

wileyj committed:
Merge branch 'next' into feat/pox-4-signing-keys

jferrant committed:
Make tx fee a variable in stackerdb dkg test and increase it

Signed-off-by: Jacinta Ferrant <jacinta@trustmachines.co>

xoloki committed:
reduce number of signers and keys so test doesn't consume all available file descriptors

obycode committed:
fix: handle epochs 2.5 and 3.0 correctly

Commit:
fix: RPC endpoints should default to proof=1

Commit:
test: fix httpcore test for defaulting in proof query string

wileyj committed:
use bitcoin 25.0 for tests

Commit:
fix: set maxburncap to 1BTC in sendrawtransaction (new in Bitcoin 25.0)

xoloki committed:
set test_stackerdb_dkg to ignore

jferrant committed:
Remove PoisonMicroBlocks from will_admit_mempool_tx check

Signed-off-by: Jacinta Ferrant <jacinta@trustmachines.co>

jferrant committed:
Add a test to check rpc submit tx mempool admission checks pass

Signed-off-by: Jacinta Ferrant <jacinta@trustmachines.co>

jferrant committed:
Run rust fmt

Signed-off-by: Jacinta Ferrant <jacinta@trustmachines.co>

setzeus committed:
catch-up & unit test fix

setzeus committed:
rebase & removed duplicate code

setzeus committed:
Merge branch 'next' into feat/pox-4-signing-keys

setzeus committed:
data structs updated

setzeus committed:
update-current-signer-set wip

setzeus committed:
updated update-current-signer-set wip

friedger committed:
Merge branch 'next' into feat/pox-4-signing-keys

setzeus committed:
added updaet-current-signer function

setzeus committed:
added get-signer-set-by-reward-cycle func

setzeus committed:
added get-signer-key-ids-by-reward-cycle func

setzeus committed:
fixed parameter length

setzeus committed:
simplified get-signer-set-by-reward-cycle

jcnelson committed:
Merge branch 'next' into feat/pox-4-signing-keys

setzeus committed:
addressing latest comments

setzeus committed:
Merge remote-tracking branch 'refs/remotes/origin/next' into feat/pox-4-signing-keys

jferrant commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -216,6 +220,14 @@
+;; Map that tracks the signing-key & cycle to a list of key-ids
+(define-map reward-cycle-signing-key-ids 
+    {cycle: uint, signer: (buff 33)}
+    {key-ids: (list 4096 uint)})
```
I think it should be only 4000 key ids (same number of reward slots at least which I think is 4000? but maybe its 4096?)

xoloki replied:
Yeah unless we've upped the number of reward slots?

setzeus replied:
Copy. Thought I heard 4096 in one of the latest meetings but will change both to 4000.

friedger replied:
This value is different on testnet (1000) and we can't use a constant :-/

jferrant replied:
I think the 4096 is the number of slots total that can be in a given stacker-db instance. :)

setzeus replied:
Double-checking here:

1. The max list size for the signer set should be 4096*
2. The max list size for the key-ids should be 4000*

Correct?

jcnelson replied:
They need to be the same value.  I'm fine with it being 4000.

jferrant commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -1311,6 +1323,62 @@
+        ;; Assert that signer key-ids haven't been set yet (map entry is empty)
+        (asserts! (is-none (map-get? reward-cycle-signer-key-ids {cycle: next-reward-cycle, signer: signer})) (err ERR_KEY_IDS_ALREADY_SET))
+
+        ;; Q - do we need another check here that we're in the 100-block prepare phase?
```
Prob not a bad idea to have this check. The node is the only thing that would ever call this function, but in theory we would never be setting the signer set except in the prepare phase.

setzeus committed:
fixed failing unit test

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -1311,6 +1323,62 @@
+(define-private (update-current-signer-set (new-signer-set (list 4096 {signer: (buff 33)})))
+    (let 
+        (
+            (next-reward-cycle (+ u1 current-pox-reward-cycle))
```
Did you test that this code passes type-check?  Because, I'm not seeing the definition of `current-pox-reward-cycle`.

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -1311,6 +1323,62 @@
+        (
+            (next-reward-cycle (+ u1 current-pox-reward-cycle))
+            (next-reward-cycle-burn-height (reward-cycle-to-burn-height next-reward-cycle))
+            (next-signer-set-map (map-get? reward-cycle-signing-key-ids next-reward-cycle))
```
This definitely doesn't type-check -- the key type for `reward-cycle-signing-key-ids` is `{ cycle: uint, signer: (buff 33) }` not `uint`.

jcnelson reviewed (changes requested):
Hi @setzeus, please make sure the code type-checks before asking us to review it.  Thanks!

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -1311,6 +1323,62 @@
             unlock-burn-height: new-unlock-ht }))))
 
+;; This function allows a node to register a new 'current-signer-set' during the prepare phase
+(define-private (update-current-signer-set (new-signer-set (list 4096 {signer: (buff 33)})))
```
Nit: this is a one-off setter, so perhaps call it `set-current-signer-set`?

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -1311,6 +1323,62 @@
+        (asserts! (is-none next-signer-set-map) (err ERR_NEXT_SIGNER_SET_ALREADY_REGISTERED))
+
+        ;; Assert in prepare phase (in the last 100 blocks of the current cycle)
+        (asserts! 
```
Can you break this conjunction up into two `asserts!`?  It's easier to read and test that way.

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -1311,6 +1323,62 @@
+        (asserts! (is-some (index-of current-signers signer)) (err ERR_NOT_IN_SIGNER_SET))
+
+        ;; Assert that signer key-ids haven't been set yet (map entry is empty)
+        (asserts! (is-none (map-get? reward-cycle-signer-key-ids {cycle: next-reward-cycle, signer: signer})) (err ERR_KEY_IDS_ALREADY_SET))
```
Where is `reward-cycle-signer-key-ids` defined?  Did you mean `reward-cycle-signing-key-ids`?

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -1311,6 +1323,62 @@
+        ;; Q - do we need another check here that we're in the 100-block prepare phase?
+
+        ;; Update map with new key-ids for signer
+        (map-set reward-cycle-signing-key-ids {cycle: next-reward-cycle, signer: signer} {key-ids: key-idss})
```
Typo -- `key-idss`

setzeus committed:
removed unused err

setzeus committed:
updated existing pox_4 test

setzeus committed:
addressed comments

setzeus committed:
format update

moodmosaic committed:
fix: Correct typecheck in `get_current_signer_set` test

- Address type mismatch in POX 4 signer set retrieval
- Update variable and function usage for type compatibility

moodmosaic committed:
fix: Correct typecheck in `get_current_signer_set` test

- Address type mismatch in PoX-4 signer set retrieval
- Update variable and function usage for type compatibility

jcnelson commented on stackslib/src/chainstate/stacks/boot/contract_tests.rs:
```diff
@@ -559,6 +561,31 @@ impl HeadersDB for TestSimHeadersDB {
+    sim.execute_next_block(|_env| {});
+    sim.execute_next_block(|_env| {});
+
+    sim.execute_next_block(|env| {
```
Can you verify that this actually catches syntax errors in `pox-4`?  Because, it hadn't in the past.

jcnelson commented on stackslib/src/chainstate/stacks/boot/mod.rs:
```diff
@@ -34,6 +34,7 @@ use clarity::vm::types::{
 };
 use clarity::vm::{ClarityVersion, Environment, SymbolicExpression};
 use lazy_static::lazy_static;
+use libc::sem_t;
```
What's this?

jcnelson commented on stackslib/src/chainstate/stacks/boot/mod.rs:
```diff
@@ -1735,6 +1765,38 @@ pub mod test {
         make_tx(key, nonce, 0, payload)
     }
 
+    // Get the current signer set
```
Please don't commit commented-out code

jcnelson commented on stackslib/src/chainstate/stacks/boot/mod.rs:
```diff
@@ -1798,7 +1860,11 @@ pub mod test {
+            vec![
+                Value::UInt(lock_period),
+                addr_tuple,
+                Value::Optional(OptionalData { data: None }),
```
The `make_pox_4_extend()` function needs to also take an `Option<StacksPublicKey>` so it can be added into this call to `stack-extend`.

jcnelson replied:
Generally speaking, you shouldn't put in default values to PoX-4 callers like `make_pox_4_extend`.  Anything you pass into `.pox-4` should be a Rust argument.

jcnelson commented on stackslib/src/chainstate/stacks/boot/mod.rs:
```diff
@@ -1165,6 +1166,37 @@ impl StacksChainState {
     }
+
+    // Get the current signer set
+    pub fn get_current_signer_set_pox_4(
```
I don't think you currently have the means of testing this, because there isn't any logic written to write state into `reward-cycle-signing-keys`.  So, I think one of two things need to happen:

* You can add the logic to insert state into `reward-cycle-signing-keys`, e.g. via a private function in `pox-4` that the node can call.  If you take this route, then you should also add some code to actually populate this data map so you can proceed to test this function in a unit test.

* You can delete this function for now, and we'll add it back in once it's possible to populate `reward-cycle-signing-keys`.

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -206,6 +221,14 @@
+;; Map that tracks the signing-key & cycle to a list of key-ids
+(define-map reward-cycle-signing-key-ids 
+    {cycle: uint, signer: (buff 33)}
+    {key-ids: (list 4000 uint)})
```
Can you add a comment explaining what a `key-id` is supposed to be, and why it's necessary in addition to the `signer-address` in `reward-cycle-signing-keys` and `signer` in `reward-cycle-signing-key-ids`?  Also, why have both the signing key and the address?

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -576,13 +603,14 @@
 ;; Lock up some uSTX for stacking!  Note that the given amount here is in micro-STX (uSTX).
 ;; The STX will be locked for the given number of reward cycles (lock-period).
-;; This is the self-service interface.  tx-sender will be the Stacker.
+;; This is the self-service interface.  tx-sender will be the Stacker. Stacker is responsible for providing a valid signing key.
```
Please word-wrap, and also it's `The Stacker is responsible ...`.

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -576,13 +603,14 @@
 ;; at the time this method is called.
 ;; * You may need to increase the amount of uSTX locked up later, since the minimum uSTX threshold
 ;; may increase between reward cycles.
+;; * You need to provide a signing-key used to allocate signature-weight to a valid signer
```
`signature weight`

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -1275,12 +1313,72 @@
+        (asserts! (is-none next-signer-set-map) (err ERR_NEXT_SIGNER_SET_ALREADY_REGISTERED))
+
+        ;; Assert in prepare phase (height equal to / greater than start of prepare phase)
+        (asserts! (>= block-height (- next-reward-cycle-burn-height u100)) (err ERR_NOT_IN_PREPARE_PHASE_EARLY))
```
This should be a `define-read-only` function called `is-in-prepare-phase`.  I think one of @friedger's PRs has this?

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:
```diff
@@ -67,6 +67,7 @@ use crate::chainstate::stacks::index::marf::MarfConnection;
 use crate::chainstate::stacks::index::MarfTrieId;
 use crate::chainstate::stacks::tests::make_coinbase;
 use crate::chainstate::stacks::*;
+use crate::chainstate::{self};
```
This is a nonsensical import

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:
```diff
@@ -1234,3 +1235,36 @@ fn balances_from_keys(
+        let reward_cycle = burnchain
+            .block_height_to_reward_cycle(tip.block_height)
+            .unwrap() as u64;
+        peer.chainstate()
```
This doesn't test anything.  There's nothing in this data map, so there's nothing to check. Please actually add data to the data map and verify that you can read it out.

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -1275,12 +1313,72 @@
             unlock-burn-height: new-unlock-ht }))))
 
+;; This function allows a node to register a new 'current-signer-set' during the prepare phase
+(define-private (set-current-signer-set (new-signer-set (list 4000 {signer: (buff 33)})))
```
There is currently no test coverage for this.  There should be a unit test in `pox_4_tests.rs` that calls this function with some synthetic data, and reads it back out again.

jcnelson replied:
Your tests should hit not only the happy path, but also each of the failure paths (and verify that the right error code is returned).

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -1275,12 +1313,72 @@
+
+;; This function allows a node to register a list of key-ids to the current-signer-set
+;; This work is done one-at-a-time by the node & could require up to 4000 individual writes
+(define-private (set-current-signer (signer (buff 33)) (key-ids (list 4000 uint)))
```
There is currently no test coverage for this.  Please add it to `pox_4_tests.rs`.  As before, test each error path as well as the happy path.

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -1275,12 +1313,72 @@
+;; Gets the signer-set for a given reward cycle
+(define-read-only (get-signer-set-by-reward-cycle (cycle uint))
+    (let ((cycle-to-burn-height (reward-cycle-to-burn-height cycle))
+            (burn-height-to-id-header-hash (get-block-info? id-header-hash cycle-to-burn-height)))
```
This doesn't type-check, since `get-block-info?` returns an `(optional ..)`.  Again, please type-check your code before asking for a review. Use `clarity-cli check` or have your unit tests run an analysis pass.

jcnelson reviewed (changes requested):
Hey @setzeus, there's still lots of work to do here.  In particular, it seems that your unit tests don't verify that `pox-4` even type-checks.  Please fix this, and please add the unit tests described in the above comments.  Thanks!

friedger commented:
Should be re-based to next or closed.

setzeus commented:
@friedger waiting for @MarvinJanssen to complete his .Signers work to see which of the maps created in this PR will stay here (aka will re-base off of his branch once published or will close).

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
